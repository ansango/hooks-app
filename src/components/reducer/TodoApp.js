import { useReducer } from "react";
import { todoReducer } from "./todoReducer";

const generateId = () => new Date().getTime().toString();

const initialState = [
  {
    id: generateId(),
    description: "Learn React",
    done: false,
  },
];
export const TodoApp = () => {
  const [todos, dispatchTodo] = useReducer(todoReducer, initialState);

  const handleSubmit = (e) => {
    e.preventDefault();
    const newTodo = { id: generateId(), description: "new todo", done: false };
    const action = {
      type: "add",
      payload: newTodo,
    };
    dispatchTodo(action);
  };
  return (
    <>
      <p>useReducer sample:</p>
      <br />

      <div className="row">
        <div className="col-12 col-md-7">
          <h3> Todos: ({todos.length})</h3>
          <hr />
          <ul className="list-group list-group-flush">
            {todos.map((todo, index) => (
              <li
                className="list-group-item d-flex items-center justify-content-between"
                key={todo.id}
              >
                <p>
                  {index + 1} - {todo.description}
                </p>
                <button className="btn btn-danger">Delete</button>
              </li>
            ))}
          </ul>
        </div>
        <div className="col-12 col-md-5">
          <h3> Add Todo</h3>
          <hr />
          <form onSubmit={handleSubmit} className="py-2">
            <input
              type="text"
              placeholder="Add a todo"
              name="description"
              autoComplete="off"
              className="form-control mb-3"
            />
            <button className="btn btn-outline-primary w-100" type="submit">
              Add
            </button>
          </form>
        </div>
      </div>
    </>
  );
};
