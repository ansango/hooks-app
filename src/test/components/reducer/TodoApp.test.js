import { act } from "@testing-library/react";
import { mount, shallow } from "enzyme";
import { TodoApp } from "../../../components/reducer/TodoApp";
import { todos } from "../../fixtures/todos";

describe("Test <TodoApp/>", () => {
  const wrapper = shallow(<TodoApp />);
  Storage.prototype.setItem = jest.fn(() => {});
  test("should render", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("should add a todo", () => {
    const wrapper = mount(<TodoApp />);
    act(() => {
      wrapper.find("TodoAdd").prop("handleAddTodo")(todos[0]);
      wrapper.find("TodoAdd").prop("handleAddTodo")(todos[1]);
      wrapper.find("TodoAdd").prop("handleAddTodo")(todos[2]);
    });

    expect(wrapper.find("h3.title").text().trim()).toBe("Todos: (3)");
    expect(localStorage.setItem).toHaveBeenCalledTimes(2);
  });

  test("should delete a todo", () => {
    wrapper.find("TodoAdd").prop("handleAddTodo")(todos[0]);
    wrapper.find("TodoList").prop("handleDeleteTodo")(todos[0].id);
    expect(wrapper.find("h3.title").text().trim()).toBe("Todos: (0)");
  });
});
