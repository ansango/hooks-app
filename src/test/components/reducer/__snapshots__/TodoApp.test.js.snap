// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <TodoApp/> should render 1`] = `
<Fragment>
  <p>
    useReducer sample:
  </p>
  <br />
  <div
    className="row"
  >
    <div
      className="col-12 col-md-7"
    >
      <h3
        className="title"
      >
        Todos: (
        0
        )
      </h3>
      <hr />
      <TodoList
        handleDeleteTodo={[Function]}
        handleToggleTodo={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col-12 col-md-5"
    >
      <TodoAdd
        handleAddTodo={[Function]}
      />
    </div>
  </div>
</Fragment>
`;
